File: src/inspectah/architect/loader.py
"""Load refined fleet tarballs for architect analysis."""

from __future__ import annotations

import json
import logging
import tarfile
from pathlib import Path

from inspectah.architect.analyzer import FleetInput

logger = logging.getLogger(__name__)


def load_refined_fleets(input_dir: Path) -> list[FleetInput]:
    """Load refined fleet tarballs from a directory.

    Each tarball should contain an inspection-snapshot.json with fleet metadata.
    Returns a list of FleetInput objects ready for the analyzer.
    """
    fleets: list[FleetInput] = []

    if not input_dir.exists():
        return fleets

    for path in sorted(input_dir.iterdir()):
        if not (path.suffix == ".gz" and path.name.endswith(".tar.gz")):
            continue

        try:
            snapshot = _extract_snapshot(path)
        except Exception as e:
            logger.warning("Skipping %s: %s", path.name, e)
            continue

        if snapshot is None:
            logger.warning("No inspection-snapshot.json found in %s", path.name)
            continue

        fleet_input = _snapshot_to_fleet_input(snapshot)
        fleets.append(fleet_input)

    return fleets


def _extract_snapshot(tarball_path: Path) -> dict | None:
    """Extract inspection-snapshot.json from a tarball."""
    with tarfile.open(tarball_path, "r:gz") as tar:
        for member in tar.getmembers():
            if member.name.endswith("inspection-snapshot.json"):
                f = tar.extractfile(member)
                if f is None:
                    continue
                return json.loads(f.read())
    return None


def _snapshot_to_fleet_input(snapshot: dict) -> FleetInput:
    """Convert an inspection snapshot dict to a FleetInput."""
    meta = snapshot.get("meta", {})
    hostname = meta.get("hostname", "unknown")
    fleet_meta = meta.get("fleet", {})
    host_count = fleet_meta.get("total_hosts", 1)

    # Extract package NVRAs
    rpm = snapshot.get("rpm", {})
    packages = [
        pkg.get("nvra", pkg.get("name", ""))
        for pkg in rpm.get("packages_added", [])
        if pkg.get("nvra") or pkg.get("name")
    ]

    # Extract config file paths
    config = snapshot.get("config", {})
    configs = [f.get("path", "") for f in config.get("files", []) if f.get("path")]

    base_image = rpm.get("base_image", "")

    # Extract preflight data
    preflight = snapshot.get("preflight", {})
    unavailable_packages = list(preflight.get("unavailable", [])) if preflight else []
    direct_install_packages = list(preflight.get("direct_install", [])) if preflight else []
    unverifiable = preflight.get("unverifiable", []) if preflight else []
    unverifiable_packages = [uv.get("name", "") for uv in unverifiable if uv.get("name")]
    preflight_status = preflight.get("status", "skipped") if preflight else "skipped"

    return FleetInput(
        name=hostname,
        packages=packages,
        configs=configs,
        host_count=host_count,
        base_image=base_image if isinstance(base_image, str) else "",
        unavailable_packages=unavailable_packages,
        direct_install_packages=direct_install_packages,
        unverifiable_packages=unverifiable_packages,
        preflight_status=preflight_status,
    )

File: cmd/inspectah/internal/architect/loader.go
package architect

import (
	"archive/tar"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// snapshotJSON is the minimal snapshot structure needed by the architect
// loader. We only parse the fields we actually need rather than importing
// the full schema package.
type snapshotJSON struct {
	Meta      map[string]interface{} `json:"meta"`
	OsRelease *struct {
		VersionID string `json:"version_id"`
	} `json:"os_release"`
	Rpm *struct {
		PackagesAdded []struct {
			Name string `json:"name"`
			NVRA string `json:"nvra"`
		} `json:"packages_added"`
		BaseImage *string `json:"base_image"`
	} `json:"rpm"`
	Config *struct {
		Files []struct {
			Path string `json:"path"`
		} `json:"files"`
	} `json:"config"`
	Preflight *struct {
		Unavailable   []string `json:"unavailable"`
		DirectInstall []string `json:"direct_install"`
		Unverifiable  []string `json:"unverifiable"`
		Status        string   `json:"status"`
	} `json:"preflight"`
}

// LoadRefinedFleets loads fleet tarballs from a directory and returns
// FleetInput objects ready for the analyzer.
//
// Each tarball should contain an inspection-snapshot.json with fleet
// metadata. Non-tarball files are skipped.
func LoadRefinedFleets(inputDir string) ([]FleetInput, error) {
	entries, err := os.ReadDir(inputDir)
	if err != nil {
		return nil, fmt.Errorf("read input directory: %w", err)
	}

	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Name() < entries[j].Name()
	})

	var fleets []FleetInput
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		if !strings.HasSuffix(name, ".tar.gz") {
			continue
		}

		path := filepath.Join(inputDir, name)
		snap, err := extractSnapshot(path)
		if err != nil {
			log.Printf("warning: skipping %s: %v", name, err)
			continue
		}
		if snap == nil {
			log.Printf("warning: no inspection-snapshot.json in %s", name)
			continue
		}

		fi := snapshotToFleetInput(snap)
		fleets = append(fleets, fi)
	}

	return fleets, nil
}

// ValidateFleetVersions enforces that all fleets share the same OS major
// version. Returns an error if versions are mixed.
func ValidateFleetVersions(fleets []FleetInput, inputDir string) error {
	if len(fleets) < 2 {
		return nil
	}

	// Reload snapshots to check version_id
	entries, err := os.ReadDir(inputDir)
	if err != nil {
		return nil
	}

	var versions []string
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".tar.gz") {
			continue
		}
		path := filepath.Join(inputDir, entry.Name())
		snap, err := extractSnapshot(path)
		if err != nil || snap == nil {
			continue
		}
		if snap.OsRelease != nil && snap.OsRelease.VersionID != "" {
			versions = append(versions, snap.OsRelease.VersionID)
		}
	}

	if len(versions) < 2 {
		return nil
	}

	// Extract major versions (e.g., "9" from "9.4")
	majors := make(map[string]bool)
	for _, v := range versions {
		major := v
		if idx := strings.Index(v, "."); idx > 0 {
			major = v[:idx]
		}
		majors[major] = true
	}

	if len(majors) > 1 {
		ms := make([]string, 0, len(majors))
		for m := range majors {
			ms = append(ms, m)
		}
		sort.Strings(ms)
		return fmt.Errorf("mixed OS major versions: %v (architect requires all fleets to share the same major version)", ms)
	}

	return nil
}

func extractSnapshot(tarballPath string) (*snapshotJSON, error) {
	f, err := os.Open(tarballPath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	gz, err := gzip.NewReader(f)
	if err != nil {
		return nil, fmt.Errorf("gzip: %w", err)
	}
	defer gz.Close()

	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("tar: %w", err)
		}
		if strings.HasSuffix(hdr.Name, "inspection-snapshot.json") {
			data, err := io.ReadAll(tr)
			if err != nil {
				return nil, fmt.Errorf("read snapshot: %w", err)
			}
			var snap snapshotJSON
			if err := json.Unmarshal(data, &snap); err != nil {
				return nil, fmt.Errorf("parse snapshot: %w", err)
			}
			return &snap, nil
		}
	}
	return nil, nil
}

func snapshotToFleetInput(snap *snapshotJSON) FleetInput {
	hostname := "unknown"
	if h, ok := snap.Meta["hostname"].(string); ok {
		hostname = h
	}

	hostCount := 1
	if fleet, ok := snap.Meta["fleet"].(map[string]interface{}); ok {
		if th, ok := fleet["total_hosts"].(float64); ok {
			hostCount = int(th)
		}
	}

	var packages []string
	if snap.Rpm != nil {
		for _, pkg := range snap.Rpm.PackagesAdded {
			nvra := pkg.NVRA
			if nvra == "" {
				nvra = pkg.Name
			}
			if nvra != "" {
				packages = append(packages, nvra)
			}
		}
	}

	var configs []string
	if snap.Config != nil {
		for _, f := range snap.Config.Files {
			if f.Path != "" {
				configs = append(configs, f.Path)
			}
		}
	}

	baseImage := ""
	if snap.Rpm != nil && snap.Rpm.BaseImage != nil {
		baseImage = *snap.Rpm.BaseImage
	}

	var unavailable, directInstall, unverifiable []string
	preflightStatus := "skipped"
	if snap.Preflight != nil {
		unavailable = snap.Preflight.Unavailable
		directInstall = snap.Preflight.DirectInstall
		unverifiable = snap.Preflight.Unverifiable
		if snap.Preflight.Status != "" {
			preflightStatus = snap.Preflight.Status
		}
	}

	return FleetInput{
		Name:                  hostname,
		Packages:              packages,
		Configs:               configs,
		HostCount:             hostCount,
		BaseImage:             baseImage,
		UnavailablePackages:   unavailable,
		DirectInstallPackages: directInstall,
		UnverifiablePackages:  unverifiable,
		PreflightStatus:       preflightStatus,
	}
}
